src/communication/communication_channel.cpp

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (C) 2008 Search Solution Corporation. All rights reserved by Search Solution.
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU General Public License as published by
+ *   the Free Software Foundation; either version 2 of the License, or
+ *   (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ */
+
+/*
+ * communication_channel.cpp - wrapper for communication primitives
+ *                           - it can establish a connection, send/receive messages, wait for events and close the connection
+ */
+
+#include "communication_channel.hpp"
+
+#include "connection_support.h"
+#include "connection_globals.h"
+#include "system_parameter.h"
+#include "thread_manager.hpp"
+#include <string>
+#include "connection_sr.h"
+
+#if defined(WINDOWS)
+#include "wintcp.h"
+#else /* WINDOWS */
+#include "tcp.h"
+#endif /* WINDOWS */
+
+communication_channel::communication_channel (int max_timeout_in_ms) :
+  m_max_timeout_in_ms (max_timeout_in_ms),
+  m_type (CHANNEL_TYPE::NO_TYPE),
+  m_socket (INVALID_SOCKET)
+{
+}
+
+communication_channel::communication_channel (communication_channel &&comm) : m_max_timeout_in_ms (
+	  comm.m_max_timeout_in_ms)
+{
+  m_type = comm.m_type;
+  comm.m_type = NO_TYPE;
+
+  m_socket = comm.m_socket;
+  comm.m_socket = INVALID_SOCKET;
+}
+
+communication_channel &communication_channel::operator= (communication_channel &&comm)
+{
+  assert (!is_connection_alive ());
+  this->~communication_channel ();
+
+  new (this) communication_channel (std::forward <communication_channel> (comm));
+  return *this;
+}
+
+communication_channel::~communication_channel ()
+{
+  close_connection ();
+}
+
+int communication_channel::send (const std::string &message)
+{
+  return communication_channel::send (message.c_str (), message.length ());
+}
+
+int communication_channel::recv (char *buffer, int &received_length)
```

Comment:
```received_length``` has a double role here (see ```css_net_recv```):

  1. input max length 
  2. output received length
I think a suggestive name would help.

Return type should be ```css_error_code```.

It is preferable to use std::size_t for memory size argument (convert to int and back internally).

```c
css_error_code communication_channel::recv (char *buffer, std::size_t & maxlen_in_recvlen_out)
```